### PR TITLE
feat: passive round-robin health check

### PR DESCRIPTION
**Pull request by shuaijinchao**
Link: apache/apisix#5158

### CODE REVIEW

**Comment by shuaijinchao** [comment]
CI failure reason: #147

**Comment by spacewander** [comment]
Let's merge master to make CI pass.

**Comment by shuaijinchao** [comment]
@spacewander After Pull #147 is merged into the current PR, the CI running cannot be triggered. What should be done?

File: lib/resty/etcd/v3.lua [hunk 1]
```diff
@@ -92,16 +87,19 @@ local function http_request_uri(self, http_cli, method, uri, body, headers, keep
...
     if err then
-        if health_check.conf ~= nil then
+        if health_check.conf.shm_name ~= nil then
             health_check.report_failure(endpoint.http_host)
-            err = endpoint.http_host .. ": " .. err
+        else
+            health_check.report_round_robin_target_failure(endpoint.http_host)
         end
-        return nil, err
+        return nil, endpoint.http_host .. ": " .. err
```
Comment (tzssangglass) [inline lib/resty/etcd/v3.lua:95]: can we move `if ... else ...` to health_check?

File: lib/resty/etcd/v3.lua [hunk 2]
```diff
@@ -45,24 +44,20 @@ local refresh_jwt_token
...
-        for _, endpoint in ipairs(endpoints) do
+    for _, endpoint in ipairs(endpoints) do
+        if health_check.conf.shm_name ~= nil then
             if health_check.get_target_status(endpoint.http_host) then
                 return endpoint
             end
+        else
+            if health_check.get_round_robin_target_status(endpoint.http_host) then
+                return endpoint
+            end
```
Comment (tzssangglass) [inline lib/resty/etcd/v3.lua:56]: can we move if ... else ... to health_check?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -56,7 +56,7 @@ jobs:
           git clone https://github.com/openresty/test-nginx.git test-nginx
           cd test-nginx && (sudo cpanm --notest . > build.log 2>&1 || (cat build.log && exit 1)) && cd ..
 
-          wget https://raw.githubusercontent.com/api7/apisix-build-tools/master/build-apisix-openresty.sh
+          wget https://raw.githubusercontent.com/api7/apisix-build-tools/v2.2.1/build-apisix-openresty.sh
```
[the remark this answers is not in the document]
Comment (shuaijinchao) [reply .github/workflows/ci.yml:59]: Please check the above record, this change was pulled in the master branch.

File: lib/resty/etcd/v3.lua [hunk 4]
```diff
@@ -710,10 +714,12 @@ local function request_chunk(self, method, path, opts, timeout)
         if not body then
             return nil, "failed to decode json body: " .. (err or " unkwon")
         elseif body.error and body.error.http_code >= 500 then
-            if health_check.conf ~= nil then
+            if health_check.conf.shm_name ~= nil then
                 -- health_check retry should do nothing here
                 -- and let connection closed to create a new one
                 health_check.report_failure(endpoint.http_host)
+            else
+                health_check.report_round_robin_target_failure(endpoint.http_host)
```
Comment (nic-chen) [inline lib/resty/etcd/v3.lua:722]: We could move all the codes related to the health check and can be moved to `health_check` to `health_check`

File: lib/resty/etcd/v3.lua (hunk 1 above)
Comment (shuaijinchao) [reply lib/resty/etcd/v3.lua]: updated

**Comment by shuaijinchao** [comment]
@tzssangglass @nic-chen please review

**Review by tzssangglass — APPROVED** [review]

File: lib/resty/etcd/v3.lua [hunk 5]
```diff
@@ -269,6 +253,10 @@ function _M.new(opts)
         return nil, err
     end
 
+    if health_check.conf == nil then
+        health_check.init({})
```
Comment (spacewander) [inline lib/resty/etcd/v3.lua:257]: Look like it makes the health check always enabled?

File: health_check.md [hunk 6]
```diff
@@ -32,7 +32,7 @@ Get the current status of the target.
 
 | name         | Type    | Requirement | Default | Description                                                  |
 | ------------ | ------- | ----------- | ------- | ------------------------------------------------------------ |
-| shm_name     | string  | required    |         | the declarative `lua_shared_dict` is used to store the health status of endpoints. |
+| shm_name     | string  | optional    |         | the declarative `lua_shared_dict` is used to store the health status of endpoints, if this option is not set, the health check will return to round-robin check mode. |
```
Comment (spacewander) [inline health_check.md:35]: Should explain what `round-robin check mode` is

Comment (shuaijinchao) [reply health_check.md]: updated

**Review by spacewander — APPROVED** [review]

**Pull request merged**